File: codes/matlab/ch20/rK1_composed_map.m
function rK1_composed_map()
%% rK1_composed_map - Chapter 20, Etude 20.10.  Global expansion of r K_1(r).
% Composed map  r = arcsinh(exp y)  + TB_n basis on (-infty, +infty).
% Author: Dr. Denys Dutykh.

    script_dir = fileparts(mfilename('fullpath'));
    addpath(fullfile(script_dir, '..', 'ch07'));
    out_dir = fullfile(script_dir, '..', '..', '..', ...
                       'textbook', 'figures', 'ch20', 'matlab');
    if ~exist(out_dir, 'dir'); mkdir(out_dir); end

    Ns = [8 12 16 20 24 32 48 64];
    L = 4;
    r_fine = logspace(-3, 1.6, 400);
    truth = r_fine .* besselk(1, r_fine);
    errs = zeros(size(Ns));
    for i = 1:length(Ns)
        a = tbn_approx(Ns(i), L);
        approx = evaluate(a, r_fine, L);
        errs(i) = max(abs(approx - truth));
    end

    NAVY=[20 45 110]/255; CORAL=[231 76 60]/255; TEAL=[22 160 133]/255;
    fig = figure('Position',[100 100 1020 340],'Color','w');

    subplot(1,2,1);
    semilogx(r_fine, truth, 'Color', NAVY, 'LineWidth',1.2); hold on;
    a16 = tbn_approx(16, L);
    semilogx(r_fine, evaluate(a16, r_fine, L), '--', 'Color', CORAL);
    xlabel('r'); ylabel('f(r)'); grid on; box on;
    title('(a) r K_1(r) (N=16 approx)');

    subplot(1,2,2);
    semilogy(Ns, errs + 1e-18, '-o', 'Color', TEAL);
    xlabel('N'); ylabel('max error'); grid on; box on;
    title(sprintf('(b) Subgeometric descent, L=%d', L));

    print(fig, fullfile(out_dir, 'rK1_composed_map.pdf'), '-dpdf');
    print(fig, fullfile(out_dir, 'rK1_composed_map.png'), '-dpng');
    fprintf('[20.10-matlab] figure saved\n');
end

function r = r_of_y(y)
    r = asinh(exp(y));
end

function y = y_of_r(r)
    y = log(sinh(r));
end

function a = tbn_approx(N, L)
    [~, x] = cheb_matrix(N);
    interior = abs(x) < 1 - 1e-12;
    y = zeros(size(x));
    y(interior) = L * x(interior) ./ sqrt(1 - x(interior).^2);
    r = r_of_y(y);
    fv = zeros(size(x));
    fv(interior) = r(interior) .* besselk(1, r(interior));
    % reorder: cheb_matrix returns x descending, so x(1) = 1 (y = +infty), x(end) = -1 (y = -infty)
    fv(end) = 1;         % r -> 0, r K_1(r) -> 1
    fv(1) = 0;           % r -> +infty, r K_1(r) -> 0
    fv = fv(:);
    V = [fv; fv(N:-1:2)];
    A = real(fft(V))/N; A(1) = A(1)/2; A(N+1) = A(N+1)/2;
    a = A(1:N+1);
end

function v = evaluate(a, r, L)
    y = y_of_r(r);
    x = y ./ sqrt(L^2 + y.^2);
    T0 = ones(size(r)); T1 = x;
    v = a(1)*T0 + a(2)*T1;
    for n = 2:length(a)-1
        Tk = 2*x.*T1 - T0;
        v = v + a(n+1)*Tk;
        T0 = T1; T1 = Tk;
    end
end

File: codes/julia/ch20/rK1_composed_map.jl
# rK1_composed_map.jl
# Chapter 20: Spectral Methods on Unbounded Intervals
# Computational Etude 20.10: r K_1(r) via arcsinh(exp y) composed map.
#
# Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)

using CairoMakie
using FFTW
using Printf
using SpecialFunctions

include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))

r_of_y(y) = asinh(exp(y))
y_of_r(r) = log(sinh(r))
target_rK1(r) = r * besselk(1, r)

function dct1_coeffs(v)
    N = length(v) - 1
    V = vcat(v, reverse(v[2:N]))
    A = real.(fft(V)) ./ N
    A[1] *= 0.5; A[N + 1] *= 0.5
    return A[1:N + 1]
end

function cheb_eval(a, x, N)
    T0 = ones(length(x)); T1 = copy(x)
    val = a[1] .* T0 .+ a[2] .* T1
    for n in 2:N
        Tk = 2 .* x .* T1 .- T0
        val .+= a[n + 1] .* Tk
        T0 = T1; T1 = Tk
    end
    return val
end

function tbn_approx(N, ell)
    _, x = cheb_matrix(N)
    fv = zeros(length(x))
    ok = abs.(x) .< 1.0 - 1e-12
    y_ok = ell .* x[ok] ./ sqrt.(1 .- x[ok] .^ 2)
    r_ok = r_of_y.(y_ok)
    fv[ok] .= target_rK1.(r_ok)
    # cheb_matrix returns x descending: x[1] = 1 (y = +infty -> r K_1 -> 0),
    # x[end] = -1 (y = -infty -> r -> 0 -> r K_1 -> 1).
    fv[1] = 0.0
    fv[end] = 1.0
    return dct1_coeffs(fv)
end

function evaluate(a, r, ell)
    y = y_of_r.(r)
    x = y ./ sqrt.(ell^2 .+ y .^ 2)
    return cheb_eval(a, x, length(a) - 1)
end

function run()
    outdir = joinpath(@__DIR__, "..", "..", "..",
                      "textbook", "figures", "ch20", "julia")
    mkpath(outdir)
    ell = 4.0
    r_fine = exp10.(range(-3.0, 1.6, length=400))
    truth = target_rK1.(r_fine)
    Ns = [8, 12, 16, 20, 24, 32, 48, 64]
    errs = [maximum(abs.(evaluate(tbn_approx(N, ell), r_fine, ell) .- truth)) for N in Ns]

    NAVY = colorant"#142D6E"; CORAL = colorant"#E74C3C"; TEAL = colorant"#16A085"
    fig = Figure(size=(1020, 340))

    ax1 = Axis(fig[1, 1], xlabel="r", ylabel="f(r)", xscale=log10,
               title="(a) r K_1(r) and N=16 approximation")
    lines!(ax1, r_fine, truth; color=NAVY, linewidth=1.2, label="exact")
    lines!(ax1, r_fine, evaluate(tbn_approx(16, ell), r_fine, ell);
           color=CORAL, linestyle=:dash, label="N=16 TB_n")
    axislegend(ax1; position=:rt)

    ax2 = Axis(fig[1, 2], xlabel="N", ylabel="max error",
               yscale=log10, title="(b) Subgeometric descent, ell=$ell")
    scatterlines!(ax2, Ns, max.(errs, 1e-18); color=TEAL)

    save(joinpath(outdir, "rK1_composed_map.pdf"), fig)
    save(joinpath(outdir, "rK1_composed_map.png"), fig)
    @printf("[20.10-julia] saved figure\n")
end

if abspath(PROGRAM_FILE) == @__FILE__
    run()
end
